include/swift/Syntax/SyntaxParsingContext.h

```diff
@@ -15,146 +15,184 @@
-
-  SyntaxParsingContextChild(SyntaxParsingContext *&ContextHolder,
-    SyntaxKind KnownSyntax): SyntaxParsingContextChild(ContextHolder,
-                             None, KnownSyntax) {};
-
-  SyntaxParsingContextChild(SyntaxParsingContext *&ContextHolder,
-                            bool Disable = false);
-
-  ~SyntaxParsingContextChild();
-  void makeNode(SyntaxKind Kind, SourceLoc LastTokLoc) override;
-  void finalize() override;
-  SyntaxParsingContext* getParent() { return Parent; }
-  void setSyntaxKind(SyntaxKind SKind) override;
-  void setContextKind(SyntaxContextKind CKind) override;
-  SyntaxParsingContextRoot &getRoot();
-  SyntaxParsingContextKind getKind() override {
-    return SyntaxParsingContextKind::Child;
-  };
+  /// Construct root context.
+  SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder, SourceFile &SF);
+
+  /// Designated constructor for child context.
+  SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder)
+      : Parent(CtxtHolder), CtxtHolder(CtxtHolder), 
+        Enabled(Parent->isEnabled()) {
+    assert(CtxtHolder->isTopOfContextStack() &&
+           "SyntaxParsingContext cannot have multiple children");
+    CtxtHolder = this;
+  }
+
+  SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder, SyntaxContextKind Kind)
+      : SyntaxParsingContext(CtxtHolder) {
+    setCoerceKind(Kind);
+  }
+
+  SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder, SyntaxKind Kind)
+      : SyntaxParsingContext(CtxtHolder) {
+    setCreateSyntax(Kind);
+  }
+
+  ~SyntaxParsingContext();
+
+  void disable() { Enabled = false; }
+  bool isEnabled() const { return Enabled; }
+  bool isRoot() const { return !Parent; }
+  bool isTopOfContextStack() const { return this == CtxtHolder; }
+
+  SyntaxParsingContext *getRoot();
+
+  /// Add RawSyntax to the parts.
+  void addRawSyntax(RC<RawSyntax> Raw);
+
+  /// Add Token with Trivia to the parts.
+  void addToken(Token &Tok, Trivia &LeadingTrivia, Trivia &TrailingTrivia);
+
+  /// Add Syntax to the parts.
+  void addSyntax(Syntax Node);
+
+  RC<RawSyntax> popBack() {
+    auto Raw = std::move(Parts.back());
+    Parts.pop_back();
+    return Raw;
+  }
+
+  template<typename SyntaxNode>
+  llvm::Optional<SyntaxNode> popIf() {
+    if (auto Node = make<Syntax>(Parts.back()).getAs<SyntaxNode>()) {
+      Parts.pop_back();
+      return Node;
+    }
+    return None;
+  }
+
+  TokenSyntax popToken() {
+    assert(Parts.back()->Kind == SyntaxKind::Token);
+    return make<TokenSyntax>(popBack());
+  }
+
+  /// Create a syntax node using the tail \c N elements of collected parts and
+  /// replace those parts with the single result.
+  void createNodePartially(SyntaxKind Kind, size_t N);
+
+  /// Create a node using the tail of the collected parts. The number of parts
+  /// is automatically determined from \c Kind. Node: limited number of \c Kind
+  /// are supported. See the implementation.
+  void createNodePartially(SyntaxKind Kind);
```

Comment:
This name seems to be wrong considering what the function does. Could we rename both these two functions to "createNodeInPlace()"